query/functions.go

```diff
@@ -2150,3 +2156,110 @@ func (r *UnsignedBottomReducer) Emit() []UnsignedPoint {
 	sort.Sort(sort.Reverse(&h))
 	return points
 }
+
+type StringMergeHllReducer struct {
+	plus *hll.Plus
+	err  error
+}
+
+func NewStringMergeHllReducer() *StringMergeHllReducer {
+	return &StringMergeHllReducer{plus: nil}
+}
+
+func unmarshalPlus(s string) (*hll.Plus, error) {
+	if string(hllPrefix) != s[:len(hllPrefix)] {
+		if string(hllErrorPrefix) == s[:len(hllErrorPrefix)] {
+			// parse a special error out of the string.
+			return nil, fmt.Errorf("%v", s[len(hllErrorPrefix):])
+		}
+		return nil, fmt.Errorf("Bad prefix for hll.Plus")
+	}
+	data := []byte(s[len(hllPrefix):])
+	if len(data) == 0 {
+		// explicitly treat as empty no-op
+		return nil, nil
+	}
+	b := make([]byte, base64.StdEncoding.DecodedLen(len(data)))
+	_, _ = base64.StdEncoding.Decode(b, data)
+	h := new(hll.Plus)
+	if err := h.UnmarshalBinary(b); err != nil {
+		return nil, err
+	}
+	return h, nil
+}
+
+func (r *StringMergeHllReducer) AggregateString(p *StringPoint) {
+	// we cannot return an error because returning an error slows all aggregation
+	// functions by ~1%. So we hack around it by marshalling the error as a string.
+	if r.err != nil {
+		return
+	}
+	h, err := unmarshalPlus(p.Value)
+	if err != nil {
+		r.err = err
+		return
+	}
+	if r.plus == nil {
+		r.plus = h
+		return
+	}
+	err = r.plus.Merge(h)
+	if err != nil {
+		r.err = err
+		return
+	}
+}
+
+func marshalPlus(p *hll.Plus, err error) StringPoint {
+	if err != nil {
+		return StringPoint{
+			Time:  ZeroTime,
+			Value: string(hllErrorPrefix) + err.Error(),
+		}
+	}
+	if p == nil {
+		return StringPoint{
+			Time:  ZeroTime,
+			Value: string(hllPrefix),
+		}
+	}
+	b, _ := p.MarshalBinary() // no possible errors
```

Comment:
If a function has an error return value, it's best to check it even if the current implementation of the function doesn't currently return an error. If someone adds an error return in the future, they're unlikely check every caller and shouldn't have to since the contract says it's okay to return an error.

Reply:
Good point. This is left over from when I was hoping not to return any error at all - fixed.